src/main/scala/net/snowflake/spark/snowflake/io/CloudStorageOperations.scala

```diff
@@ -537,15 +544,30 @@ sealed trait CloudStorage {
   : SingleElementIterator = {
     val fileName = getFileName(partitionID, format, compress)
 
-    // Upload data when reading data if upload retry is disabled.
-    // It can save some memory for writing to snowflake.
-    // NOTE: For GCP, the data have to be cached and then uploaded.
-    val disable_cache_data = maxRetryCount <= 1 && storageInfo.isDefined
+    // When the upload task fails for non-fatal error, this task will be
+    // rescheduled by spark until the max attempt number is arrived.
+    // In some case, the task failure is caused by cloud service throttling.
+    // The default sleep time in spark (3 seconds) is not enough,
+    // so we do exponential backoff here.
+    val attemptNumber = TaskContext.get().attemptNumber()
+    var backoffInfo = "no_backoff"
+    if (attemptNumber > 0) {
```

Comment:
Add the sleep at the end if the failure instead of the beginning of the second.

Reply:
OK.

Reply:
When I make the change, I found may it is better to check in the beginning. If we put the sleep in the end, when all retry fail, the last retry also needs to sleep. This sleep is useless.
It looks there is no way to get the MAX attempt number, if there is, we can avoid the last sleep.

Reply:
The possibility to hit the extra sleep is small. It's OK to put it in the end.

Reply:
When is uploadRDD used vs upload ?
It might be cleaner to remove this sleep logic from uploadPartition -- it's kind of confusing to refer to taskcontext etc. in there. 
My suggestion: revert uploadPartition to roughly what it was as of this commit: 3bcdc53. That is let's leave the sleep/retry logic out of there.
Create a function wrapping uploadPartition that's passed in a task context and sleeps based on that. And lets make sure it's called from uploadRDD and upload

Reply:
uploadRDD is used for AWS and Azure. upload is used for GCP.

Reply:
uploadPartition() has implemented the retry logic from spark 2.7.0 for GCP support. From spark 2.7.1, we refactor uploadRDD() to use uploadPartition() to avoid duplicate code. After this refactor, AWS and Azure also uses uploadPartition.
I will do following as you suggested. It's sounds clearer.
1. Introduce a function called doUploadPartition() to read and upload data.
2. uploadPartition() call doUploadPartition() with try-catch and handle the sleep/telemetry message sending, etc